webhook/server.js

```diff
@@ -0,0 +1,377 @@
+                event.payload.room_ids.forEach(async function (roomId) {
+                    // todo implement me
+                })
+            }
+        }
+    }
+
+    async getChatKitRoom(roomID, user) {
+        const cachedRoom = this.roomsCache.get(roomID)
+        if (cachedRoom) {
+            return cachedRoom
+        }
+        const room = await this.chatKitClient().getRoom({
+            userId: user,
+            roomId: roomID,
+        })
+        this.roomsCache.set(roomID, room)
+        return room
+    }
+
+    async ensureMembersExists(ids) {
+
+        const missingIds = []
+        const parent = this
+
+        for (let i = 0; i < ids.length; i++) {
+            const user = this.usersCache.get(ids[i])
+            if (!user) {
+                missingIds.push(ids[i])
+            }
+        }
+
+        // get missing users from cache
+        if (missingIds.length > 0) {
+            const loadedUsers = await this.chatKitClient().getUsersById({
+                userIds: missingIds
+            })
+
+            const users = []
+            loadedUsers.forEach(function (u) {
+                users.push(parent.toStreamUser(u))
+            })
+            await this.streamClient().updateUsers(users)
+
+            users.forEach(function (u) {
+                // update LRU cache
+                parent.usersCache.set(u.id, u)
+            })
+        }
+    }
+
+    // sanitizeUserId transform a chatKit user id to a suitable stream user id
+    sanitizeUserId(id) {
+        // todo improve me
+        return id.replace(':', '_')
+    }
+
+    // converts a chatKit user to a stream user
+    async createStreamUser(chatKitUser) {
+        await this.streamClient().updateUser(this.toStreamUser(chatKitUser))
+    }
+
+    // converts a chatKit user user to a a stream user
+    toStreamUser(chatKitUser) {
+        return {
+            id: this.sanitizeUserId(chatKitUser.id),
+            image: chatKitUser.avatar_url,// todo omit keys with empty values
+            name: chatKitUser.name,
+            ...chatKitUser.custom_data,
+        };
+    }
+
+    // converts a chatKit room to a stream channel
+    async getOrCreateRoom(room) {
+        let type = 'livestream'
+        if (room.private) {
+            type = 'messaging'
+        }
+
+        await this.ensureMembersExists(room.member_user_ids)
+        const streamChannel = this.streamClient().channel(type, room.id, {
+            name: room.name,
+            members: room.member_user_ids,
+            created_by_id: room.created_by_id,
+            ...room.custom_data
+        })
+        await streamChannel.create()
+        return streamChannel
+    }
+
+    // converts a chatKit room to a stream channel
+    async toStreamMessage(streamChannel, chatKitMessage) {
+        const streamMessage = {
+            id: chatKitMessage.id.toString(),
+            user_id: chatKitMessage.user_id,
+            text: chatKitMessage.text,
```

Comment:
```
{
  "payload": {
    "message": {
      "created_at": "2019-10-11T22:26:10Z",
      "id": ...,
      "parts": [
        {
          "content": "Hey",
          "type": "text/plain"
        }
      ],
      "room_id": "...",
      "truncated": false,
      "updated_at": "2019-10-11T22:26:10Z",
      "user_id": "..."
    },
    "offline_user_ids": [
      "...",
      "..."
    ],
    "room": {
      "created_at": "2019-10-11T15:35:02Z",
      "created_by_id": "...",
      "custom_data": {
      },
      "id": "...",
      "name": "...",
      "private": true,
      "updated_at": "2019-10-11T16:59:28Z"
    },
    "sender": {
      "created_at": "2019-09-19T08:10:36Z",
      "id": "...",
      "name": "...",
      "updated_at": "2019-09-19T08:10:36Z"
    }
  },
  "metadata": {
    "created_at": "2019-10-11T22:26:14Z",
    "event_type": "v1.message_sent_user_offline",
    "event_id": "...",
    "product": "chatkit"
  }
}
```
Have you tested this out since I don't think there's a text field on the event message. I think the text is the content for the part that has type text/plain.

Reply:
yes, this is tested.
the go client as a text field. that's why I've kept the field.
if you follow the code you will notice that the message.text field is overwritten [here](https://github.com/GetStream/pusher-chatkit-migration/pull/6/files#diff-3b3cc7d20f462dc356c060c6f8a08643R262)

Reply:
Yep, missed that!